### PR TITLE
Add Retry to instanceProfileAddRole to handle IAM's eventual consistency

### LINKED ISSUE #2012: Should terraform retry on failure when setting the EC2 instance profile?
_This issue was originally opened by @koikonom as hashicorp/terraform#16422. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Terraform version: Terraform v0.10.3

When trying to create an IAM role and attach it to existing instances at the same time we may hit an AWS API issue where the role we just created isn't yet propagated and the association attempt results in an error like this one:
```
InvalidParameterValue: Value (ConsulServer) for parameter iamInstanceProfile.name is invalid. Invalid IAM Instance Profile name
	status code: 400, request id: <request id>
```

Running the same plan for a second time was successful, with no other changes on AWS' side. I believe that if terraform retried after a small timeout like it does for other operations would have worked just fine. 

Plan:
```
  + module.instance-profiles.aws_iam_instance_profile.consul-server
      arn:         "<computed>"
      create_date: "<computed>"
      name:        "ConsulServer"
      path:        "/"
      role:        "ConsulServer"
      roles.#:     "<computed>"
      unique_id:   "<computed>"

  + module.instance-profiles.aws_iam_role.consul-server
      arn:                   "<computed>"
      assume_role_policy:    "<some json>"
      create_date:           "<computed>"
      force_detach_policies: "false"
      name:                  "ConsulServer"
      path:                  "/"
      unique_id:             "<computed>"

  + module.instance-profiles.aws_iam_role_policy.consul-server
      name:   "ConsulServer"
      policy: "<some more json>"
      role:   "ConsulServer"

  ~ module.consul.aws_instance.consul-green[0]
      iam_instance_profile: "EC2CloudWatchLogsPut" => "ConsulServer"

  ~ module.consul.aws_instance.consul-green[1]
      iam_instance_profile: "EC2CloudWatchLogsPut" => "ConsulServer"

  ~ module.consul.aws_instance.consul-green[2]
      iam_instance_profile: "EC2CloudWatchLogsPut" => "ConsulServer"
```

### Expected Behavior
IAM role should be created and attached to the instance.

### Actual Behavior
IAM role was created, but was not attached to the instance:
```
module.instance-profiles.aws_iam_role.consul-server: Creating...
  arn:                   "" => "<computed>"
  assume_role_policy:    "" => "<policy JSON goes here>"
  create_date:           "" => "<computed>"
  force_detach_policies: "" => "false"
  name:                  "" => "ConsulServer"
  path:                  "" => "/"
  unique_id:             "" => "<computed>"
module.instance-profiles.aws_iam_role.consul-server: Creation complete after 1s (ID: ConsulServer)
module.instance-profiles.aws_iam_role_policy.consul-server: Creating...
  name:   "" => "ConsulServer"
  policy: "" => "<policy JSON goes here>"
  role:   "" => "ConsulServer"
module.instance-profiles.aws_iam_instance_profile.consul-server: Creating...
  arn:         "" => "<computed>"
  create_date: "" => "<computed>"
  name:        "" => "ConsulServer"
  path:        "" => "/"
  role:        "" => "ConsulServer"
  roles.#:     "" => "<computed>"
  unique_id:   "" => "<computed>"
module.instance-profiles.aws_iam_role_policy.consul-server: Creation complete after 0s (ID: ConsulServer:ConsulServer)
module.instance-profiles.aws_iam_instance_profile.consul-server: Creation complete after 1s (ID: ConsulServer)
module.consul.aws_instance.consul-green.0: Modifying... (ID: <instance ID>)
  iam_instance_profile: "EC2CloudWatchLogsPut" => "ConsulServer"
module.consul.aws_instance.consul-green.2: Modifying... (ID: <instance ID>)
  iam_instance_profile: "EC2CloudWatchLogsPut" => "ConsulServer"
module.consul.aws_instance.consul-green.1: Modifying... (ID: <instance ID>)
  iam_instance_profile: "EC2CloudWatchLogsPut" => "ConsulServer"
Error applying plan:

3 error(s) occurred:

* module.consul.aws_instance.consul-green[1]: 1 error(s) occurred:

* aws_instance.consul-green.1: InvalidParameterValue: Value (ConsulServer) for parameter iamInstanceProfile.name is invalid. Invalid IAM Instance Profile name
	status code: 400, request id: 0ae39d73-4aff-425f-bceb-4e4da2dcbcd4
* module.consul.aws_instance.consul-green[2]: 1 error(s) occurred:

* aws_instance.consul-green.2: InvalidParameterValue: Value (ConsulServer) for parameter iamInstanceProfile.name is invalid. Invalid IAM Instance Profile name
	status code: 400, request id: aed3c0f2-8f79-474c-9825-fea313668acd
* module.consul.aws_instance.consul-green[0]: 1 error(s) occurred:

* aws_instance.consul-green.0: InvalidParameterValue: Value (ConsulServer) for parameter iamInstanceProfile.name is invalid. Invalid IAM Instance Profile name
	status code: 400, request id: e9e8eeea-dec3-494b-80bd-fb263af598b0
```

### PR DESCRIPTION
**Pull request by joakiboxxx**
Fixes #2012

### CODE REVIEW

File: aws/resource_aws_iam_instance_profile.go [hunk 1]
```diff
@@ -160,7 +160,16 @@ func instanceProfileAddRole(iamconn *iam.IAM, profileName, roleName string) erro
 		RoleName:            aws.String(roleName),
 	}
 
-	_, err := iamconn.AddRoleToInstanceProfile(request)
+	err := resource.Retry(30*time.Second, func() *resource.RetryError {
+		var err error
+		_, err = iamconn.AddRoleToInstanceProfile(request)
+		// Todo: handle retryable and non-retryable errors
```
Comment (bflad) [inline aws/resource_aws_iam_instance_profile.go]: Hi @joakiboxxx! Thanks for submitting this! Can you please update this to the below (including the comment):

```go
// IAM unfortunately does not provide a better error code or message for eventual consistency
// InvalidParameterValue: Value (XXX) for parameter iamInstanceProfile.name is invalid. Invalid IAM Instance Profile name
if isAWSErr(err, "InvalidParameterValue", "Invalid IAM Instance Profile name") {
  return resource.RetryableError(err)
}
if err != nil {
  return resource.NonRetryableError(err)
}
return nil
```

Comment (joakiboxxx) [reply aws/resource_aws_iam_instance_profile.go]: Hi Brian!

I applied your patch and tested it, and the change no longer solves my problem ;) With a debug inside the Retry callback, I get this from err:

[DEBUG] Error adding role NoSuchEntity: The role with name joakiboxxx_role cannot be found.

Wouldn't you agree that NoSuchEntity is indeed a retryable error? If so, may I propose the following:

if isAWSErr(err, "InvalidParameterValue", "Invalid IAM Instance Profile name") || isAWSErr(err, "NoSuchEntity", "The role with name") {
  return resource.RetryableError(err)
}

**Comment by joakiboxxx** [comment]
Thanks, @bflad! Happy to help :)

**Comment by bflad** [comment]
This has been released in [version 1.40.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1400-october-10-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**